### PR TITLE
Unreal: Build shot structure from the database

### PR DESCRIPTION
**Pull request by simonebarbieri**
## Changelog Description
Added a button to build the shot structure as sequences from the database.

PR linked to https://github.com/ynput/OpenPype/pull/6168

### CODE REVIEW

File: client/ayon_core/hosts/unreal/api/pipeline.py [hunk 1]
```diff
@@ -590,21 +590,21 @@ def set_sequence_hierarchy(
         hid_section.set_level_names(maps)
 
 
-def generate_sequence(h, h_dir):
+def generate_sequence(name, hierarchy_dir):
     tools = unreal.AssetToolsHelpers().get_asset_tools()
```
Comment (BigRoy) [inline client/ayon_core/hosts/unreal/api/pipeline.py:594]: Can we get a docstring for this function?

File: client/ayon_core/hosts/unreal/api/hierarchy.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    sequence_path = settings['unreal']['sequence_path']
+    if sequence_path.endswith("/"):
+        sequence_path = sequence_path.rstrip("/")
+
+    return f"/Game/{sequence_path}"
+
+
+def _create_sequence(
+        element, sequence_path, parent_path="", parent_seq=None,
+        parent_fr=None
```
Comment (BigRoy) [inline client/ayon_core/hosts/unreal/api/hierarchy.py:23]: Can we use just the full argument names for these, `parent_frame_range` is much clearer?

File: client/ayon_core/hosts/unreal/api/hierarchy.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    set_sequence_hierarchy,
+)
+
+
+def get_default_sequence_path(settings):
+    """Get default render folder from blender settings."""
+
+    sequence_path = settings['unreal']['sequence_path']
+    if sequence_path.endswith("/"):
+        sequence_path = sequence_path.rstrip("/")
```
Comment (BigRoy) [inline client/ayon_core/hosts/unreal/api/hierarchy.py:16]:
```suggestion
    sequence_path = sequence_path.rstrip("/")
```

File: client/ayon_core/hosts/unreal/api/hierarchy.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    sequence_root_name = "shots"
+
+    hierarchy = get_folders_hierarchy(project_name=project)["hierarchy"]
+
+    # Find the sequence root element in the hierarchy
+    sequence_root = next((
+        element
+        for element in hierarchy
+        if element["name"] == sequence_root_name
+    ), None)
```
Comment (BigRoy) [inline client/ayon_core/hosts/unreal/api/hierarchy.py:72]: Is the requirement of there being a `shots` folder in the hierarchy an odd requirement maybe? Seems a bit hardcoded?

**Comment by simonebarbieri** [comment]
@BigRoy sorry, I forgot to set the PR as draft! It needs some more work! I will address all of of your points when I get back to work on this :) thanks as always for the inputs on these :D

File: client/ayon_core/hosts/unreal/api/hierarchy.py (hunk 4 above)
Comment (simonebarbieri) [reply client/ayon_core/hosts/unreal/api/hierarchy.py:71]: So, yes, this is just a placeholder for now to get the functionality working, but there was discussion about creating a UI ad hoc to select which folder you want to create the shot structure for, so it's WIP :)

File: server_addon/unreal/server/version.py [hunk 5]
```diff
@@ -1 +1 @@
-__version__ = "0.1.0"
```
Comment (iLLiCiTiT) [inline server_addon/unreal/server/version.py:1]: If you don't change settings you don't need to bump version of unreal.

File: client/ayon_core/hosts/unreal/api/hierarchy.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+def build_sequence_hierarchy():
+    """
+    Builds the sequence hierarchy by creating sequences from the root element.
+
+    Raises:
+        ValueError: If the sequence root element is not found in the hierarchy.
+    """
+    print("Building sequence hierarchy...")
+
+    project = os.environ.get("AVALON_PROJECT")
```
Comment (iLLiCiTiT) [inline client/ayon_core/hosts/unreal/api/hierarchy.py:90]: Requires import `from ayon_core.pipeline import get_current_project_name`
```suggestion
    project = get_current_project_name()
```

File: server_addon/unreal/server/version.py (hunk 5 above)
Comment (simonebarbieri) [reply server_addon/unreal/server/version.py:1]: I'm adding two settings though! Or should I increase the version only if I change existing ones?

Comment (iLLiCiTiT) [reply server_addon/unreal/server/version.py:1]: Sorry missed them..

**Review by antirotor — COMMENTED** [review]
I need to go through it properly, but just quick first notes:

`set_sequence_hierarchy()` needs imho even better docstring. Or better names for arguments :) ... or both?

Like this usage:

```python
set_sequence_hierarchy(
    parents_sequence[i], parents_sequence[i + 1],
    parents_frame_range[i][1],
    parents_frame_range[i + 1][0], parents_frame_range[i + 1][1],
    [level])
)
```
I must admit I really don't know what is going on there.

Another point - since paths in UE are paths, we can use `pathlib.Path` on them so no need of `.split("/")` and stuff.

Also, some part are rather static, like `get_default_sequence_path(settings)`. And I realize we tend to do it a lot in Unreal - expecting that something is always on specific path. I am glad that some of those paths can be now set in the settings - but we might later on (not in this PR) consider using actual anatomy templates there.

File: client/ayon_core/hosts/unreal/api/hierarchy.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    style
+)
+from ayon_core.pipeline import get_current_project_name
+from ayon_core.settings import get_project_settings
+from ayon_core.tools.utils import (
+    show_message_dialog,
+    PlaceholderLineEdit,
+    SquareButton,
+)
+from ayon_core.tools.ayon_utils.widgets import (
```
Comment (iLLiCiTiT) [inline client/ayon_core/hosts/unreal/api/hierarchy.py:16]: NOTE: This has changed to `from ayon_core.tools.utils import SimpleFoldersWidget` from this PR https://github.com/ynput/ayon-core/pull/195

**Comment by antirotor** [comment]
when https://github.com/ynput/ayon-core/pull/614 is merged, we need to reopen it in https://github.com/ynput/ayon-unreal

**Comment by antirotor** [comment]
PR moved to https://github.com/ynput/ayon-unreal/pull/4

**Pull request closed without merging**